virttest/graphical_console.py

```diff
@@ -0,0 +1,225 @@
+        """
+        Key Release up.
+
+        :param keystroke: released key, it happened behind press key.
+                          if combination key was pressed, like "Shift_L-a",
+                          will release key "a" first then release key "Shift_L".
+        :param interval: the interval between two keys, default interval=0.
+        """
+        for key in self._get_seq(keystroke, interval):
+            self._key_event(self.KEY_UP, key)
+
+    def key_tap(self, keystroke, hold_time=0, interval=0):
+        """
+        Key Press down first, then Key Release up.
+
+        :param keystroke: tap key, can be a single key or a combination key.
+                          type is str, single key like 'a'.
+                          combination key connect them with '-', like "Shift_L-a".
+        :param hold_time: the time that keep one key pressed, default hole_time=0.
+        :param interval: the interval between two keys, default interval=0.
+
+        """
+        self.key_press(keystroke, interval)
+        self._hold_time(hold_time)
+        keystroke = self._reverse(keystroke)
+        self.key_release(keystroke, interval)
+
+    def _hold_time(self, hold_time):
+        """
+        Time that keep one key pressed.
+        """
+        time.sleep(hold_time)
+
+    @staticmethod
+    def _get_seq(seq_str, interval):
+        """
+        Get sequence of seq_str.
+
+        :param seq_str: key str, if have '-', seprate them by '-' sign,
+                        and return them separately. if have not '-',
+                        return orginal str.
+        :param interval: the interval between return two value.
+        """
+        seq_str += _STR_DASH
+        while True:
+            val, seq_str = seq_str.split(_STR_DASH, 1)
+            yield val
+            if not seq_str.rstrip(_STR_DASH):
+                break
+            if interval:
+                time.sleep(interval)
+
+    @staticmethod
+    def _reverse(seq_str):
+        """
+        Reversed seq_str.
+        """
+        seq = seq_str.split(_STR_DASH)
+        seq.reverse()
+        return _STR_DASH.join(seq)
+
+    def _key_convert(self, key):
+        """
+        Key Convert.
+
+        The method convert key name that defined in KEY_MAP_FILE.
+        file to correct console excuteble code.
+
+        :param key: key name.
+        """
+        raise NotImplementedError()
+
+    def _key_event(self, event, key):
+        """
+        Send Key Event.
+
+        :param event: key event.
+        :param key: key name.
+        """
+        raise NotImplementedError()
+
+    def close(self):
+        pass
+
+
+class DummyConsole(BaseConsole):
+
+    """
+    Dummy console
+    """
+    pass
+
+
+class QMPConsole(BaseConsole):
+
+    """
+    QMP console
+    """
+    KEY_DOWN = True
+    KEY_UP = False
+
+    KEY_MAP_FILE = "keymap_to_qcode.json"
+    KEY_MAP = os.path.join(data_dir.DEPS_DIR, KEY_MAP_FILE)
```

Comment:
```python
os.path.join(data_dir.get_deps_dir("$keymap_dir"), KEY_MAP_FILE)
```

Reply:
Hi xuhan, 

I cannot get the correct path with data_dir.get_deps_dir("$keymap_dir"), tested report error:  ERROR: Could not find specified deps dir after looking 10 parent directories

I checked the test result shown that get_deps_dir will search start from /usr/local/lib/python3.6/site-packages/avocado_plugins_vt-70.0-py3.6.egg/avocado_vt/,  graphical_console.py in this directory, but share/deps/keymap_dir did not in this path, after run #python3 setup.py install, it will be installed to  /usr/share/avocado-plugins-vt/shared/deps/, it's different path with graphical_console.py.  that's why I wasn't use data_dir.get_deps_dir("$keymap_dir") before.

How about I change  to KEY_MAP_FILE = "keymap_dir/keymap_to_qcode.json"?

Thanks
Peixiu